syntax = "proto3";

package university;

enum LoanStatus {
  ONGOING = 0;
  RETURNED = 1;
  LATE = 2;
}

// Book mesajı
message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 pageCount = 6;
  int32 stock = 7;
}

// Student mesajı
message Student {
  string id = 1;
  string name = 2;
  string studentNumber = 3;
  string email = 4;
  bool isActive = 5;
}

// Loan mesajı
message Loan {
  string id = 1;
  string studentId = 2;
  string bookId = 3;
  string loanDate = 4;
  string returnDate = 5;
  LoanStatus status = 6;
}

// Ortak mesajlar
message IdRequest {
  string id = 1;
}

message Empty {}

message BookList {
  repeated Book books = 1;
}

message StudentList {
  repeated Student students = 1;
}

message LoanList {
  repeated Loan loans = 1;
}

service BookService {
  rpc ListBooks(Empty) returns (BookList);
  rpc GetBook(IdRequest) returns (Book);
  rpc CreateBook(Book) returns (Book);
  rpc UpdateBook(Book) returns (Book);
  rpc DeleteBook(IdRequest) returns (Empty);
}

service StudentService {
  rpc ListStudents(Empty) returns (StudentList);
  rpc GetStudent(IdRequest) returns (Student);
  rpc CreateStudent(Student) returns (Student);
  rpc UpdateStudent(Student) returns (Student);
  rpc DeleteStudent(IdRequest) returns (Empty);
}

service LoanService {
  rpc ListLoans(Empty) returns (LoanList);
  rpc GetLoan(IdRequest) returns (Loan);
  rpc CreateLoan(Loan) returns (Loan);
  rpc ReturnLoan(IdRequest) returns (Loan);
}
